Common/File/VFS/ZipFileReader.cpp

```diff
@@ -132,26 +132,29 @@ bool ZipFileReader::GetFileListing(const char *orig_path, std::vector<File::File
 
 void ZipFileReader::GetZipListings(const char *path, std::set<std::string> &files, std::set<std::string> &directories) {
 	size_t pathlen = strlen(path);
-	if (path[pathlen - 1] == '/')
-		pathlen--;
+	if (pathlen == 1 && path[0] == '/') {
+		// Root. We simply use a zero length string.
+		pathlen = 0;
+	}
 
 	std::lock_guard<std::mutex> guard(lock_);
 	int numFiles = zip_get_num_files(zip_file_);
 	for (int i = 0; i < numFiles; i++) {
 		const char* name = zip_get_name(zip_file_, i, 0);
 		if (!name)
-			continue;
+			continue;  // shouldn't happen, I think
 		if (!memcmp(name, path, pathlen)) {
-			// The prefix is right. Let's see if this is a file or path.
 			const char *slashPos = strchr(name + pathlen + 1, '/');
 			if (slashPos != 0) {
-				// A directory.
-				std::string dirName = std::string(name + pathlen + 1, slashPos - (name + pathlen + 1));
+				// A directory. Let's pick off the only part we care about.
+				int offset = pathlen;
+				std::string dirName = std::string(name + offset, slashPos - (name + offset));
 				directories.insert(dirName);
-			} else if (name[pathlen] == '/') {
-				const char *fn = name + pathlen + 1;
+			} else {
+				// It's a file.
+				const char *fn = name + pathlen;
 				files.insert(std::string(fn));
-			}  // else, it was a file with the same prefix as the path. like langregion.ini next to lang/.
```

Comment:
How do we protect against this case now?

So let's say `path` is "assets/lang".  We did check it starts with "assets/lang", so that's good.  But "assets/langregion" matches that and does not have a /.  So won't we match it and add it to the list as "egion" or something?

I might also be missing something about `name + pathlen`.  In my above scenario, "assets/lang/foo.ini" would not find a slash after `name + pathlen + 1` (which is "foo.ini") but wouldn't `fn` become "/foo.ini"?  Seems confusing.

It feels like this new logic might only be something that works when path == ""?  Which does look like it was broken before.

-[Unknown]

Reply:
I somehow convinced myself that since all directories in a zip file always end wih '/', they'd be caught in the first case anyway - but that doesn't even make sense. Don't know what I was thinking.

I'll make a unit test for this today to really make sure all cases comes out right.

Reply:
Actually this does work - just gotta follow the convention of directory paths ending with "/" consistently throughout.

The unit test revealed a number of bugs, but I've fixed it and it's working now.

Reply:
Okay, yeah, making `GetFileListing()` add `/` makes sense.

But oops, now there's a source_assets/ziptest.zip.

-[Unknown]

Reply:
Yeah, we don't have a designated directory for testdata, at least not that I found. Though I suppose could have put the zip in /unittest too.